src/v/cluster/tests/rm_stm_compatibility_test.cc

```diff
@@ -62,14 +62,38 @@ bool compare_batch(
     return true;
 }
 
+model::record_batch make_fence_batch_v1(
+  model::producer_identity pid,
+  model::tx_seq tx_seq,
+  std::chrono::milliseconds transaction_timeout_ms) {
+    iobuf key;
+    auto pid_id = pid.id;
+    reflection::serialize(key, model::record_batch_type::tx_fence, pid_id);
+
+    iobuf value;
+    reflection::serialize(
+      value,
+      tx::fence_control_record_v1_version,
+      tx_seq,
+      transaction_timeout_ms);
+
+    storage::record_batch_builder builder(
+      model::record_batch_type::tx_fence, model::offset(0));
+    builder.set_producer_identity(pid.id, pid.epoch);
+    builder.set_control_type();
+    builder.add_raw_kv(std::move(key), std::move(value));
+
+    return std::move(builder).build();
+}
+
 SEASTAR_THREAD_TEST_CASE(fence_batch_compatibility) {
     vlog(logger.info, "Test fence_batch_v1");
```

Comment:
why do we still need it?

Reply:
we can still read old fence batch versions in the log after the upgrade. (eg: if old data is not truncated).